File: android/app/src/main/kotlin/com/hadithly/app/core/preferences/PreferencesStore.kt
package com.hadithly.app.core.preferences

import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.serialization.json.Json

private val storeJson = Json { ignoreUnknownKeys = true; encodeDefaults = true }

/**
 * Persists the F2 reader preference contract and performs the one-time
 * migration from the legacy single-language settings. The stored payload is
 * a versioned JSON blob under "reader.preferences.v1" in the shared
 * "hadithly" preferences file; the legacy keys ("user.preferredLanguage",
 * "reader.arabicFontSize") stay readable for the migration and keep being
 * written for backend compatibility. Mirrors the iOS PreferencesStore.
 */
class PreferencesStore(sharedPrefs: SharedPreferences) {

    data class MigrationResult(
        val preferences: ReaderPreferences,
        /** True when the seed came from the legacy language key. */
        val seededFromLegacy: Boolean,
    )

    private val prefs: SharedPreferences = sharedPrefs

    private val _preferences = MutableStateFlow(load(prefs))
    val preferences: StateFlow<ReaderPreferences> = _preferences

    /** Fired after every explicit local change (not after server adoption). */
    var onPreferencesChanged: ((ReaderPreferences) -> Unit)? = null

    companion object {
        const val STORE_KEY = "reader.preferences.v1"
        const val LEGACY_LANGUAGE_KEY = "user.preferredLanguage"
        const val LEGACY_ARABIC_FONT_SIZE_KEY = "reader.arabicFontSize"

        /**
         * One-time migration: an existing v1 payload wins (a later explicit
         * choice is never reset); otherwise both locales seed from the legacy
         * translation-language setting. With neither source, defaults to en/en.
         */
        fun migrate(
            storedJson: String?,
            legacyLanguage: String?,
            legacyArabicFontSize: Float? = null,
        ): MigrationResult {
            if (storedJson != null) {
                runCatching { storeJson.decodeFromString<ReaderPreferences>(storedJson) }
                    .getOrNull()
                    ?.let { stored ->
                        return MigrationResult(preferences = stored, seededFromLegacy = false)
                    }
            }
            val canonical = legacyLanguage?.let { LocaleFallback.canonicalizeLocaleTag(it) }
            if (canonical != null) {
                var seeded = ReaderPreferences(
                    uiLocale = canonical,
                    translationLocale = canonical,
                )
                if (legacyArabicFontSize != null) {
                    seeded = seeded.copy(
                        arabicFontSize = legacyArabicFontSize
                            .coerceIn(ReaderPreferences.MIN_ARABIC_FONT_SIZE, ReaderPreferences.MAX_ARABIC_FONT_SIZE),
                    )
                }
                return MigrationResult(preferences = seeded, seededFromLegacy = true)
            }
            return MigrationResult(preferences = ReaderPreferences(), seededFromLegacy = false)
        }
    }

    // Mutations

    /** Onboarding "Choose your language": writes both locales at once. */
    fun applyOnboardingLanguage(code: String) {
        val canonical = LocaleFallback.canonicalizeLocaleTag(code) ?: return
        update { it.copy(uiLocale = canonical, translationLocale = canonical) }
        writeLegacyLanguage(canonical)
    }

    fun setUILocale(code: String) {
        val canonical = LocaleFallback.canonicalizeLocaleTag(code) ?: return
        update { it.copy(uiLocale = canonical) }
    }

    fun setTranslationLocale(code: String) {
        val canonical = LocaleFallback.canonicalizeLocaleTag(code) ?: return
        update { it.copy(translationLocale = canonical) }
        writeLegacyLanguage(canonical)
    }

    fun setReadingDirection(direction: ReadingDirection) =
        update { it.copy(readingDirection = direction) }

    fun setTheme(theme: ReaderTheme) = update { it.copy(theme = theme) }

    fun setArabicFont(font: ArabicFont) = update { it.copy(arabicFont = font) }

    fun setArabicFontSize(size: Float) = update {
        it.copy(
            arabicFontSize = size.coerceIn(
                ReaderPreferences.MIN_ARABIC_FONT_SIZE,
                ReaderPreferences.MAX_ARABIC_FONT_SIZE,
            ),
        )
    }

    /**
     * Rejects the all-hidden state: returns false without a preference
     * update when both layers would be hidden.
     */
    fun setVisibility(arabic: Boolean? = null, translation: Boolean? = null): Boolean {
        val current = _preferences.value
        val nextArabic = arabic ?: current.arabicVisible
        val nextTranslation = translation ?: current.translationVisible
        if (!nextArabic && !nextTranslation) return false
        update { it.copy(arabicVisible = nextArabic, translationVisible = nextTranslation) }
        return true
    }

    /**
     * Adopts server-side preferences only when they are newer than the local
     * state (last-write-wins by updatedAt). Does not re-emit
     * onPreferencesChanged (no sync loop).
     */
    fun adoptServer(incoming: ReaderPreferences) {
        val validated = runCatching { ReaderPreferencesValidator.validated(incoming) }.getOrNull() ?: return
        if (validated.updatedAt <= _preferences.value.updatedAt) return
        _preferences.value = validated
        persist()
        writeLegacyLanguage(validated.translationLocale)
    }

    // Internals

    private fun load(target: SharedPreferences): ReaderPreferences {
        val storedJson = target.getString(STORE_KEY, null)
        val legacyLanguage = target.getString(LEGACY_LANGUAGE_KEY, null)
        val legacySize = if (target.contains(LEGACY_ARABIC_FONT_SIZE_KEY)) {
            target.getFloat(LEGACY_ARABIC_FONT_SIZE_KEY, 26f)
        } else {
            null
        }
        val result = migrate(storedJson, legacyLanguage, legacySize)
        target.edit().putString(STORE_KEY, storeJson.encodeToString(result.preferences)).apply()
        return result.preferences
    }

    private fun update(mutate: (ReaderPreferences) -> ReaderPreferences) {
        val next = mutate(_preferences.value)
            .copy(updatedAt = System.currentTimeMillis().toDouble())
        val validated = runCatching { ReaderPreferencesValidator.validated(next) }.getOrElse { return }
        _preferences.value = validated
        persist()
        onPreferencesChanged?.invoke(validated)
    }

    private fun persist() {
        prefs.edit().putString(STORE_KEY, storeJson.encodeToString(_preferences.value)).apply()
    }

    private fun writeLegacyLanguage(code: String) {
        prefs.edit().putString(LEGACY_LANGUAGE_KEY, code).apply()
    }
}

File: ios/Hadithly/Core/Preferences/PreferencesStore.swift
import Foundation
import Observation

/// Persists the F2 reader preference contract and performs the one-time
/// migration from the legacy single-language settings. The stored payload is
/// a versioned JSON blob under "reader.preferences.v1"; the legacy keys
/// ("user.preferredLanguage", "reader.arabicFontSize") stay readable for the
/// migration and keep being written for backend compatibility.
@MainActor
@Observable
final class PreferencesStore {
    static let storeKey = "reader.preferences.v1"
    static let legacyLanguageKey = "user.preferredLanguage"
    static let legacyArabicFontSizeKey = "reader.arabicFontSize"

    struct MigrationResult: Equatable {
        let preferences: ReaderPreferences
        /// True when the seed came from the legacy language key.
        let seededFromLegacy: Bool
    }

    private(set) var preferences: ReaderPreferences

    /// Fired after every explicit local change (not after server adoption).
    var onPreferencesChanged: ((ReaderPreferences) -> Void)?

    private let userDefaults: UserDefaults

    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
        let legacyLanguage = userDefaults.string(forKey: Self.legacyLanguageKey)
        let legacySize = userDefaults.object(forKey: Self.legacyArabicFontSizeKey) as? Double
        let result = Self.migrate(
            storedJSON: userDefaults.string(forKey: Self.storeKey),
            legacyLanguage: legacyLanguage,
            legacyArabicFontSize: legacySize
        )
        self.preferences = result.preferences
        userDefaults.set(
            (try? JSONEncoder().encode(result.preferences)).map { String(data: $0, encoding: .utf8) } ?? "",
            forKey: Self.storeKey
        )
    }

    /// One-time migration: an existing v1 payload wins (a later explicit
    /// choice is never reset); otherwise both locales seed from the legacy
    /// translation-language setting. With neither source, defaults to en/en.
    nonisolated static func migrate(
        storedJSON: String?,
        legacyLanguage: String?,
        legacyArabicFontSize: Double? = nil
    ) -> MigrationResult {
        if let storedJSON,
           let data = storedJSON.data(using: .utf8),
           let stored = try? JSONDecoder().decode(ReaderPreferences.self, from: data) {
            return MigrationResult(preferences: stored, seededFromLegacy: false)
        }
        if let legacyLanguage,
           let canonical = LocaleFallback.canonicalizeLocaleTag(legacyLanguage) {
            var preferences = ReaderPreferences()
            preferences.uiLocale = canonical
            preferences.translationLocale = canonical
            if let legacyArabicFontSize {
                preferences.arabicFontSize = min(
                    ReaderPreferences.maxArabicFontSize,
                    max(ReaderPreferences.minArabicFontSize, legacyArabicFontSize)
                )
            }
            return MigrationResult(preferences: preferences, seededFromLegacy: true)
        }
        return MigrationResult(preferences: ReaderPreferences(), seededFromLegacy: false)
    }

    // MARK: - Mutations

    /// Onboarding "Choose your language": writes both locales at once.
    func applyOnboardingLanguage(_ code: String) {
        guard let canonical = LocaleFallback.canonicalizeLocaleTag(code) else { return }
        update { preferences in
            preferences.uiLocale = canonical
            preferences.translationLocale = canonical
        }
        userDefaults.set(canonical, forKey: Self.legacyLanguageKey)
    }

    func setUILocale(_ code: String) {
        guard let canonical = LocaleFallback.canonicalizeLocaleTag(code) else { return }
        update { $0.uiLocale = canonical }
    }

    func setTranslationLocale(_ code: String) {
        guard let canonical = LocaleFallback.canonicalizeLocaleTag(code) else { return }
        update { $0.translationLocale = canonical }
        userDefaults.set(canonical, forKey: Self.legacyLanguageKey)
    }

    func setReadingDirection(_ direction: ReadingDirection) {
        update { $0.readingDirection = direction }
    }

    func setTheme(_ theme: ReaderTheme) {
        update { $0.theme = theme }
    }

    func setArabicFont(_ font: ArabicFont) {
        update { $0.arabicFont = font }
    }

    func setArabicFontSize(_ size: Double) {
        update {
            $0.arabicFontSize = min(
                ReaderPreferences.maxArabicFontSize,
                max(ReaderPreferences.minArabicFontSize, size)
            )
        }
    }

    /// Rejects the all-hidden state: the caller receives false instead of a
    /// preference update when both layers would be hidden.
    @discardableResult
    func setVisibility(arabic: Bool? = nil, translation: Bool? = nil) -> Bool {
        let nextArabic = arabic ?? preferences.arabicVisible
        let nextTranslation = translation ?? preferences.translationVisible
        guard nextArabic || nextTranslation else { return false }
        update {
            $0.arabicVisible = nextArabic
            $0.translationVisible = nextTranslation
        }
        return true
    }

    /// Adopts server-side preferences only when they are newer than the
    /// local state (last-write-wins by updatedAt). Does not re-emit
    /// onPreferencesChanged (no sync loop).
    func adoptServer(_ incoming: ReaderPreferences) {
        guard let validated = try? ReaderPreferencesValidator.validated(incoming) else { return }
        guard validated.updatedAt > preferences.updatedAt else { return }
        preferences = validated
        persist()
        userDefaults.set(validated.translationLocale, forKey: Self.legacyLanguageKey)
    }

    // MARK: - Internals

    private func update(_ mutate: (inout ReaderPreferences) -> Void) {
        var next = preferences
        mutate(&next)
        next.updatedAt = Date().timeIntervalSince1970 * 1000
        guard let validated = try? ReaderPreferencesValidator.validated(next) else { return }
        preferences = validated
        persist()
        onPreferencesChanged?(validated)
    }

    private func persist() {
        guard let data = try? JSONEncoder().encode(preferences) else { return }
        userDefaults.set(String(data: data, encoding: .utf8) ?? "", forKey: Self.storeKey)
    }
}
